xls/contrib/integrator/ir_integrator.cc

```diff
@@ -172,7 +237,141 @@ IntegrationFunction::UnifyIntegrationNodes(Node* node_a, Node* node_b) {
                                        /*default_value=*/std::nullopt));
 
   node_pair_to_mux_[key] = mux;
-  return UnifiedNode({mux, /*new_mux_added=*/true});
+  return UnifiedNode({mux, /*change=*/UnificationChange::kNewMuxAdded});
+}
+
+absl::StatusOr<Node*> IntegrationFunction::ReplaceMuxCases(
+    Node* mux_node,
+    const absl::flat_hash_map<int64, Node*>& source_index_to_case) {
+  XLS_RET_CHECK(IntegrationFunctionOwnsNode(mux_node));
+  XLS_RET_CHECK(!IsMappingTarget(mux_node));
+  Select* mux = mux_node->As<Select>();
+
+  // Add case.
+  std::vector<Node*> new_cases;
+  new_cases.insert(new_cases.end(), mux->cases().begin(), mux->cases().end());
+  for (const auto& index_case : source_index_to_case) {
+    XLS_RET_CHECK(IntegrationFunctionOwnsNode(index_case.second));
+    new_cases[index_case.first] = index_case.second;
+  }
+
+  // Replace mux.
+  XLS_ASSIGN_OR_RETURN(Node * new_mux,
+                       mux->ReplaceUsesWithNew<Select>(
+                           mux->selector(), new_cases,
+                           /*default_value=*/
+                           IsPowerOfTwo(new_cases.size())
+                               ? std::nullopt
+                               : absl::optional<Node*>(mux->default_value())));
+  XLS_RETURN_IF_ERROR(function()->RemoveNode(mux));
```

Comment:
Just a random note of warning. Because nodes are being deleted, xls::Node pointer values may be reused. Not sure if that is a problem or not but I've been burned by it before. A way to avoid this is to use the node ID in maps and what not rather than Node*.